lib/constructs/identity-provider.ts

```diff
@@ -0,0 +1,247 @@
+  /**
+   * The application that will function as the client.
+   */
+  clientApplication: IIdentityProviderClientApplication;
+  scopes: string[];
+}
+
+/**
+ * Properties for creating a client on the identity provider.
+ */
+export interface CognitoIdentityProviderClientProps extends IdentityProviderClientOptions {
+  /**
+   * The identity provider for which to create a client.
+   */
+  userPool: cognito.IUserPool;
+}
+
+/**
+ * A Cognito User Pool Client-based IdP Client.
+ *
+ * This wraps the User Pool Client such that the necessary attributes are made
+ * available in a generic way.
+ */
+export class CognitoIdentityProviderClient extends Construct {
+  public readonly clientId: string;
+  public readonly name: string;
+  public readonly secret: secrets.ISecret;
+
+  constructor(scope: Construct, id: string, props: CognitoIdentityProviderClientProps) {
+    super(scope, id);
+    const client = props.userPool.addClient("Client", {
+      oAuth: {
+        flows: {
+          clientCredentials: true,
+          authorizationCodeGrant: false,
+          implicitCodeGrant: false,
+        },
+        scopes: props.scopes.map(cognito.OAuthScope.custom),
+      },
+      generateSecret: true,
+    });
+    const clientDetails = new UserPoolClientDetails(this, "ClientDetails", {
+      client,
+      userPool: props.userPool,
+    });
+    this.name = clientDetails.getResponseField("UserPoolClient.ClientName");
+    this.clientId = client.userPoolClientId;
+    this.secret = new secrets.Secret(this, "ClientSecret", {
+      secretStringValue: cdk.SecretValue.resourceAttribute(
+        clientDetails.getResponseField("UserPoolClient.ClientSecret")
+      ),
+    });
+  }
+}
+
+export interface ScopeOption {
+  name: string;
+  description: string;
+}
+
+export interface ScopeConfiguration {
+  resourceServerName?: string;
+  scopes: ScopeOption[];
+}
+
+export interface IdentityProviderProps {
+  scopes?: ScopeConfiguration[];
+}
+
+/**
+ * A generic Identity Provider application.
+ */
+export interface IIdentityProvider {
+  readonly domainName: string;
+  addClient(app: IIdentityProviderClientApplication, scopes: string[]): IIdentityProviderClient;
+}
+
+/**
+ * Required properties for creating an Identity Provider.
+ */
+export interface CognitoIdentityProviderProps extends IdentityProviderProps {
+  domainPrefix?: string;
+}
+
+/**
+ * An internal identity provider for the application.
+ */
+export class CognitoIdentityProvider extends Construct implements IIdentityProvider {
+  public readonly domainName: string;
+
+  public readonly userPool: cognito.IUserPool;
+  public readonly domain: cognito.IUserPoolDomain;
+
+  private readonly resourceServers: cognito.UserPoolResourceServer[] = [];
+  constructor(scope: Construct, id: string, props?: CognitoIdentityProviderProps) {
+    super(scope, id);
+    this.userPool = new cognito.UserPool(this, "IdP", {});
+    this.domain = this.userPool.addDomain("Domain", {
```

Comment:
`domain` named Domain?
Should we name this atat-domain or something else?

Reply:
Nope! In the construct tree this will end up having the names concatenated. We probably actually need to get _out_ of the habit of jamming `Atat` in front of everything and just defining it in the top level. You can see the emitted in Template to see that it still has a unique logical ID